File: MochilaFracionaria.java
// Obtive ajuda de uma IA na transcrição do código abaixo. 
// Sua versão inicial foi feita em Python, pois é a linguagem que conheço. 
// Inclusive, disponibilizei o código original em um arquivo neste mesmo repositório.

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

class Item {
    String name;
    double weight;
    double value;
    double ratio; // valor por kg

    public Item(String name, double weight, double value) {
        this.name = name;
        this.weight = weight;
        this.value = value;
        this.ratio = value / weight;
    }
}

class FractionalKnapsackSolver {
    double capacity;
    List<Item> items;
    double totalValue;
    List<SelectedItem> selectedItems;
    double usedCapacity;

    // Classe auxiliar para armazenar o item selecionado e sua fração
    static class SelectedItem {
        String name;
        double fraction;
        double value;

        public SelectedItem(String name, double fraction, double value) {
            this.name = name;
            this.fraction = fraction;
            this.value = value;
        }
    }

    public FractionalKnapsackSolver(double capacity, List<Item> items) {
        this.capacity = capacity;
        this.items = items;
        this.totalValue = 0.0;
        this.selectedItems = new ArrayList<>();
        this.usedCapacity = 0.0;
    }

    public double solve() {
        // Ordena os itens pelo valor por kg (decrescente)
        items.sort(Comparator.comparingDouble((Item i) -> i.ratio).reversed());

        double remainingCapacity = capacity;
        System.out.printf("Capacidade total da mochila: %.2f kg%n%n", capacity);

        for (Item item : items) {
            if (remainingCapacity == 0) {
                break;
            }

            if (remainingCapacity >= item.weight) {
                // Adiciona o item completo
                remainingCapacity -= item.weight;
                totalValue += item.value;
                usedCapacity += item.weight;
                selectedItems.add(new SelectedItem(item.name, 1.0, item.value));
                System.out.printf("Adicionou o item completo: %s - %.2f kg - R$%.2f%n",
                        item.name, item.weight, item.value);
            } else {
                // Adiciona fração do item
                double fraction = remainingCapacity / item.weight;
                double addedValue = item.value * fraction;
                totalValue += addedValue;
                usedCapacity += remainingCapacity;
                selectedItems.add(new SelectedItem(item.name, fraction, addedValue));
                System.out.printf("Adicionou %.2f%% do item: %s - %.2f kg - R$%.2f%n",
                        fraction * 100, item.name, remainingCapacity, addedValue);
                remainingCapacity = 0;
            }
        }

        displaySummary();
        return totalValue;
    }

    private void displaySummary() {
        System.out.printf("%nValor total final da mochila: R$%.2f%n", totalValue);
        System.out.println("\nResumo dos itens selecionados:");
        for (SelectedItem si : selectedItems) {
            String itemType = si.fraction == 1.0 ? "Item completo" : String.format("%.2f%% do item", si.fraction * 100);
            System.out.printf("- %s: %s (Valor: R$%.2f)%n", itemType, si.name, si.value);
        }

        double filledPercentage = (usedCapacity / capacity) * 100;
        double remainingPercentage = 100 - filledPercentage;

        System.out.printf("%nA mochila foi preenchida em %.2f%%%n", filledPercentage);
        System.out.printf("Espaço restante na mochila: %.2f%%%n", remainingPercentage);
    }
}

public class MochilaFracionaria {
    public static void main(String[] args) {
        double backpackCapacity = 16.0;

        List<Item> items = new ArrayList<>();
        items.add(new Item("Ouro", 10, 600000));
        items.add(new Item("Grãos de café", 15, 30000));
        items.add(new Item("Arroz", 8, 12500));
        items.add(new Item("Farinha", 20, 10000));
        items.add(new Item("Açúcar", 7, 8000));
        items.add(new Item("Petróleo", 19, 150000));
        items.add(new Item("Cobre", 30, 90000));
        items.add(new Item("Sal", 22, 5000));

        FractionalKnapsackSolver solver = new FractionalKnapsackSolver(backpackCapacity, items);
        solver.solve();
    }
}

File: MochilaFracionaria.py
import sys
import locale

# Configuração para saída UTF-8 (evita problemas com caracteres especiais no terminal)
sys.stdout.reconfigure(encoding='utf-8')

# Tenta configurar o locale para formato de moeda brasileiro (útil se precisar formatar moeda no futuro)
try:
    locale.setlocale(locale.LC_ALL, 'pt_BR.UTF-8')
except locale.Error:
    print("Erro ao configurar o locale para pt_BR.UTF-8.")

class Item:
    # Classe que representa um item com nome, peso e valor, além do seu valor por unidade de peso (ratio)
    def __init__(self, name, weight, value):
        # name: Nome do item.
        # weight: Peso do item em kg.
        # value: Valor total do item.
        self.name = name
        self.weight = weight
        self.value = value
        self.ratio = value / weight  # Valor por kg (usado para priorizar no algoritmo)

class FractionalKnapsackSolver:
    # Classe responsável por resolver o problema da Mochila Fracionária
    def __init__(self, capacity, items):
        # capacity: Capacidade máxima da mochila (em kg).
        # items: Lista de objetos do tipo Item.
        self.capacity = capacity
        self.items = items
        self.totalValue = 0.0
        self.selectedItems = []
        self.usedCapacity = 0.0  # Para calcular quanto da mochila foi preenchido

    def solve(self):
        # Executa o algoritmo da mochila fracionária.
        # Ordena os itens por valor por kg, seleciona os itens inteiros ou frações que couberem na mochila,
        # acumula o valor total e exibe um resumo detalhado da seleção.

        # Ordena os itens pelo valor por kg (ordem decrescente)
        self.items.sort(key=lambda item: item.ratio, reverse=True)
        remainingCapacity = self.capacity

        print(f"Capacidade total da mochila: {self.capacity} kg\n")

        for item in self.items:
            if remainingCapacity == 0:
                break  # Se a mochila estiver cheia, para o processamento

            if remainingCapacity >= item.weight:
                # Caso o item inteiro caiba na mochila
                remainingCapacity -= item.weight
                self.totalValue += item.value
                self.usedCapacity += item.weight
                self.selectedItems.append((item.name, 1.0, item.value))
                print(f"Adicionou o item completo: {item.name} - {item.weight} kg - R${item.value:.2f}")
            else:
                # Caso só uma fração do item caiba na mochila
                fraction = remainingCapacity / item.weight
                addedValue = item.value * fraction
                self.totalValue += addedValue
                self.usedCapacity += remainingCapacity
                self.selectedItems.append((item.name, fraction, addedValue))
                print(f"Adicionou {fraction * 100:.2f}% do item: {item.name} - {remainingCapacity:.2f} kg - R${addedValue:.2f}")
                remainingCapacity = 0

        self.displaySummary()
        return self.totalValue

    def displaySummary(self):
        # Exibe o resumo final: valor total acumulado e detalhamento de cada item selecionado (inteiro ou parcial)
        print(f"\nValor total final da mochila: R${self.totalValue:.2f}")
        print("\nResumo dos itens selecionados:")
        for name, fraction, value in self.selectedItems:
            itemType = "Item completo" if fraction == 1.0 else f"{fraction * 100:.2f}% do item"
            print(f"- {itemType}: {name} (Valor: R${value:.2f})")

        # Calcula a porcentagem de preenchimento e espaço restante
        filledPercentage = (self.usedCapacity / self.capacity) * 100
        remainingPercentage = 100 - filledPercentage

        print(f"\nA mochila foi preenchida em {filledPercentage:.2f}%")
        print(f"Espaço restante na mochila: {remainingPercentage:.2f}%")

if __name__ == "__main__":
    # Define a capacidade máxima da mochila (em kg)
    backpackCapacity = 16  # kg

    # Lista de itens disponíveis (nome, peso, valor total)
    items = [
        Item("Ouro", 10, 600000),
        Item("Grãos de café", 15, 30000),
        Item("Arroz", 8, 12500),
        Item("Farinha", 20, 10000),
        Item("Açúcar", 7, 8000),
        Item("Petróleo", 19, 150000),
        Item("Cobre", 30, 90000),
        Item("Sal", 22, 5000)
    ]

    solver = FractionalKnapsackSolver(backpackCapacity, items)
    solver.solve()
